src/net_processing.cpp

```diff
@@ -3329,6 +3329,11 @@ void PeerManagerImpl::ProcessCompactBlockTxns(CNode& pfrom, Peer& peer, const Bl
 
         PartiallyDownloadedBlock& partialBlock = *range_flight.first->second.second->partialBlock;
 
+        if (partialBlock.header.IsNull()) {
```

Comment:
might note that the header is blanked in FillBlock, but the partial block never deleted, which is why this would catch the behavior

Reply:
Yup good point, will do.

Reply:
Done in 0df02ac0e870bc34c90f95d2a429e6dcc9217dfa

Reply:
Wondering if we should always wipe the `partialBlock` (i.e. downgrade the request to a regular block request) when `FillBlock` fails. Then it would no longer be possible to process two `blocktxn` messages for the same block.

Reply:
> where all of our other peers don't support compact block

This, or a malicious peer somehow controls all three slots (harder since one slot is outbound), or we are in `-blocksonly` mode.

> In any case, this was non-obvious to me. Maybe add a comment in the spot where we fall back to GETDATA explaining that we don't wipe the partialBlock there on purpose?

Good idea, I will add a comment. I am not sure if the existing logic is on purpose, but the way it is written in master makes it impossible to stall block download indefinitely since the first_in_flight peer needs to make forward progress after the getdata fallback after 10 minutes or get disconnected.

Reply:
It may be possible to call `reset()` on `partialBlock` to wipe it. I was assuming that wiping would mean calling `RemoveBlockRequest` (resets `m_downloading_since` which allows block download stalling) followed by `BlockRequested`. I will have to think more on this because I am not sure if it introduces a footgun.